### PR TITLE
Support opt-out of extra derived traits for rejections for `#[derive(FromRequest)]`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Makes it possible to write

```rust
#[derive(FromRequest)]
#[from_request(rejection_derive(!Display, !Error))]
struct Extractor {
    other: OtherExtractor,
}
```

Which opts out of the rejection implementing `Display` and `Error`.

The allowed syntax inside `rejection_derive` is intentionally quite limited. Can always be expanded in the future if we needed.

Fixes https://github.com/tokio-rs/axum/issues/726
Fixes https://github.com/tokio-rs/axum/issues/725

### CODE REVIEW

File: axum-macros/src/from_request/attr.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        if lh.peek(kw::via) {
+            let via = input.parse::<kw::via>()?;
+            let content;
+            syn::parenthesized!(content in input);
+            content.parse().map(|path| Self::Via { via, path })
+        } else {
+            Err(lh.error())
+        }
+    }
+}
```
Comment (davidpdrsn) [inline axum-macros/src/from_request/attr.rs:150]: The duplication in this parsing code bothers me a little but I couldn't find a nicer solution that gave good error messages.

File: axum-macros/src/from_request.rs [hunk 2]
```diff
@@ -270,7 +288,7 @@ fn extract_each_field_rejection(
         }
     };
 
-    let impl_display = {
+    let impl_display = if rejection_derive_opt_outs.derive_display() {
```
Comment (jplatte) [inline axum-macros/src/from_request.rs:291]: Could use `bool::then` (plus `.transpose()?`) here.

File: axum-macros/src/from_request/attr.rs (hunk 1 above)
Comment (jplatte) [reply axum-macros/src/from_request/attr.rs:150]: Seems fine to me as-is.

File: axum-macros/src/from_request/attr.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        let mut debug = None;
+        let mut display = None;
+        let mut error = None;
+
+        while !input.is_empty() {
+            input.parse::<Token![!]>()?;
+
+            let lh = input.lookahead1();
+            if lh.peek(kw::Debug) {
+                debug = Some(input.parse()?);
```
Comment (jplatte) [inline axum-macros/src/from_request/attr.rs:184]: Should probably raise an error on multiple assignment here too? Could avoid the extra nesting by moving the assignment or error into the helper function.

**Review by jplatte — APPROVED** [review]

File: axum-macros/src/from_request.rs (hunk 2 above)
Comment (davidpdrsn) [reply axum-macros/src/from_request.rs:291]: Hm maybe. I think its fine as is.

**Comment by davidpdrsn** [comment]
I fixed a few other things I discovered:

- Added docs for `rejection_derive`
- Multiple opt outs of the same trait like you mentioned
- Using both `via` and `rejection_derive`. Since `via` on the container changes it to use the "via extractors"'s rejection, and doesn't generate a few type, you cannot opt out of which traits are derived for it
- Better error message if opting out of `Debug` or `Display` with also opting out of `Error`. `std::error::Error` requires `Debug` and `Display` so the error message was slightly confusing

File: axum-macros/src/from_request.rs [hunk 4]
```diff
@@ -313,11 +333,19 @@ fn extract_each_field_rejection(
...
+        })
+    } else {
+        None
+    };
+
+    let impl_debug = if rejection_derive_opt_outs.derive_debug() {
+        Some(quote! { #[derive(Debug)] })
+    } else {
+        None
     };
```
Comment (jplatte) [inline axum-macros/src/from_request.rs:345]: Here you don't even need `.transpose()?` 😉

```suggestion
    let impl_debug = rejection_derive_opt_outs.derive_debug().then(|| {
        quote! { #[derive(Debug)] }
    });
```

But of course the current code is fine too.

**Review by jplatte — APPROVED** [review]

File: axum-macros/src/from_request.rs (hunk 4 above)
Comment (davidpdrsn) [reply axum-macros/src/from_request.rs]: Yeah thats definitely nicer!

**Pull request merged**